### DISCUSSION TITLE
Custom Type does not apply immediately after calling this.em.persistAndFlush

### DISCUSSION THREAD

**Discussion by winsontam** [mikro-orm/mikro-orm#4117] [General]
**Describe the bug**
I created a Custom Type for lower case for all input value as below
```
import { EntityProperty, Platform, Type } from '@mikro-orm/core';

type JSType = string | null | undefined;
type DBType = string | null | undefined;

export class LowerCaseType extends Type<JSType, DBType> {
  convertToDatabaseValue(value: JSType): DBType {
    if (typeof value === 'undefined' || value === null) {
      return value as null | undefined;
    }

    return value.toLowerCase();
  }

  convertToJSValue(value: DBType): JSType {
    if (!value) {
      return value as null | undefined;
    }

    return value.toLowerCase();
  }

  getColumnType(prop: EntityProperty, platform: Platform) {
    return platform.getVarcharTypeDeclarationSQL(prop);
  }

  compareAsType() {
    return 'string';
  }
}
```

And applied in User Email field
```
export class User {
  [EntityRepositoryType]?: UserRepository;

  @PrimaryKey({ type: 'uuid', defaultRaw: 'uuid_generate_v4()' })
  id: string;

  @Field(() => String)
  @Unique({ name: 'UserEmailUniqueConstraint' })
  @Property({ type: LowerCaseType })
  email: string;
}
```

When i try to persistAndFlush, the user email is not changed to lower case, but in DB already is lower case value
```
const user = this.userRepository.create({
  email: 'ImNotOnlyLowerCase@example.com',
});

await this.em.persistAndFlush([user]);

console.log(user.email); // ImNotOnlyLowerCase@example.com

em.clear();
const user = this.userRepository.findOne({id: user.id});
console.log(user.email); // imnotonlylowercase@example.com
```

**Expected behavior**
Should apply lower case immediately after persistAndFlush

**Additional context**
I suggest use returning with Custom Type setting.

https://github.com/mikro-orm/mikro-orm/blob/8bb0268a6d67143b0aa4dd0a5c6a6fb1bd0f8374/packages/knex/src/query/QueryBuilderHelper.ts#L591

**Versions**

| Dependency | Version |
| - | - |
| node | ? |
| typescript | ? | 
| mikro-orm | ? |
| your-driver | postgres |

**Comment by B4nan** [comment]
That's not how it works, custom types are not supposed to be rehydrated. Only the "to db value" method gets triggered on flush. The use case is rather mapping a database value to a different type (usually some value object), back and forth - not to normalize the value to the same type.

You can use `onUpdate: e => e.email.toLowerCase()` in the decorator options to enforce lowercase values instead of your custom type.

**Reply by winsontam** [reply]
oh, thanks

i make this and good for my logic
```
  @BeforeCreate()
  @BeforeUpdate()
  async lowerCaseEmail(event: EventArgs<this>) {
    if (event.changeSet.payload?.email) {
      this.email = this.email.toLowerCase();
    }
  }
```
